src/cluster_slots.c

```diff
@@ -0,0 +1,217 @@
+    clusterNode *n = clusterNodeGetMaster(server.cluster->myself);
+    return server.cluster->slots[slot] == n;
+}
+
+static void markAssignedSlots(unsigned char *slots) {
+    for (int slot = 0; slot < CLUSTER_SLOTS; slot++) {
+        if (doesSlotBelongToMyShard(slot)) slots[slot]++;
+    }
+}
+
+static int countAssignedSlotsFromSlotsArray(unsigned char *slots) {
+    int count = 0;
+    for (int slot = 0; slot < CLUSTER_SLOTS; slot++) {
+        if (slots[slot]) count++;
+    }
+    return count;
+}
+
+static void checkSlotAssignment(unsigned char *slots, int start_slot, int end_slot) {
+    for (int slot = start_slot; slot <= end_slot; slot++) {
+        if (doesSlotBelongToMyShard(slot)) {
+            slots[slot]++;
+        }
+    }
+}
+
+static uint64_t getSingleSlotStat(int slot, int order_by) {
+    serverAssert(order_by != ORDER_BY_INVALID);
+    uint64_t singleSlotStat = 0;
+    if (order_by == ORDER_BY_KEY_COUNT) {
+        singleSlotStat = countKeysInSlot(slot);
+    }
+    return singleSlotStat;
+}
+
+static int slotStatEntryAscCmp(const void *a, const void *b) {
+    sortedSlotStatEntry entry_a = *((sortedSlotStatEntry *) a);
+    sortedSlotStatEntry entry_b = *((sortedSlotStatEntry *) b);
+    return entry_a.stat - entry_b.stat;
+}
+
+static int slotStatEntryDescCmp(const void *a, const void *b) {
+    sortedSlotStatEntry entry_a = *((sortedSlotStatEntry *) a);
+    sortedSlotStatEntry entry_b = *((sortedSlotStatEntry *) b);
+    return entry_b.stat - entry_a.stat;
+}
+
+static void sortSlotStats(sortedSlotStatEntry sorted[], int order_by, int desc) {
+    int i = 0;
+    
+    for (int slot = 0; slot < CLUSTER_SLOTS; slot++) {
+        if (doesSlotBelongToMyShard(slot)) {
+            sorted[i].slot = slot;
+            sorted[i].stat = getSingleSlotStat(slot, order_by);
+            i++;
+        }
+    }
+    qsort(sorted, i, sizeof(sortedSlotStatEntry), (desc) ? slotStatEntryDescCmp : slotStatEntryAscCmp);
+}
+
+static void addReplySingleSlotStat(client *c, int slot) {
+    addReplyLongLong(c, slot);
+    addReplyMapLen(c, 1);
+    addReplyBulkCString(c, "key-count");
+    addReplyLongLong(c, countKeysInSlot(slot));
+}
+
+static void addReplySlotStats(client *c, unsigned char *slots) {
+    int num_slots_assigned = countAssignedSlotsFromSlotsArray(slots);
+    addReplyMapLen(c, num_slots_assigned);
+
+    for (int slot = 0; slot < CLUSTER_SLOTS; slot++) {
+        if (slots[slot]) addReplySingleSlotStat(c, slot);
+    }
+}
+
+static void addReplySortedSlotStats(client *c, sortedSlotStatEntry sorted[], long limit) {
+    int num_slots_assigned = getMyShardSlotCount();
+    int len = min(limit, num_slots_assigned);
+    addReplyMapLen(c, len);
+
+    for (int i = 0; i < len; i++) {
+        addReplySingleSlotStat(c, sorted[i].slot);
+    }
+}
+
+static void sortAndAddReplySlotStats(client *c, int order_by, long limit, int desc) {
+    sortedSlotStatEntry sorted[CLUSTER_SLOTS];
+    sortSlotStats(sorted, order_by, desc);
```

Comment:
Nt - this function does more than sorting.

```suggestion
    collectAndSortSlotStats(sorted, order_by, desc);
```